### PR TITLE
Create `CarouselView` widget - Part 2

### PR DESCRIPTION
**Pull request by QuncCccccc**
This PR is to create `Carousel.weighted` so the size of each carousel item is based on a list of weights. While scrolling, item sizes are changing dynamically based on the scrolling progress.

https://github.com/flutter/flutter/assets/36861262/181472b0-6f8b-48e7-b191-ab5f7c88c0c8

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

### CODE REVIEW

**Comment by QuncCccccc** [comment]
Since Hans came back, I also added him as a reviewer😁

File: packages/flutter/test/material/carousel_test.dart [hunk 1]
```diff
@@ -445,13 +687,113 @@ void main() {
...
+    expect(getItem(9), findsOneWidget);
+    expect(tester.getRect(getItem(9)), const Rect.fromLTRB(240.0, 0.0, 560.0, 600.0));
+  });
+
+  testWidgets('While scrolling, one more item will show at the end of the screen during items transition', (WidgetTester tester) async {
+    await tester.pumpWidget(
+      MaterialApp(
+        home: Scaffold(
+          body: CarouselView.weighted(
+            layoutWeights: const <int>[1,2,4,2,1],
```
Comment (goderbauer) [inline packages/flutter/test/material/carousel_test.dart:738]:
```suggestion
            layoutWeights: const <int>[1, 2, 4, 2, 1],
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 2]
```diff
@@ -4,6 +4,7 @@
 
 import 'dart:math' as math;
 
+import 'package:collection/collection.dart';
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:7]: Wondering: What are you using from this package?

File: packages/flutter/lib/src/material/carousel.dart [hunk 3]
```diff
@@ -15,31 +16,60 @@ import 'theme.dart';
 
 /// A Material Design carousel widget.
 ///
-/// The [CarouselView] present a scrollable list of items, each of which can dynamically
+/// The [CarouselView] presents a scrollable list of items, each of which can dynamically
 /// change size based on the chosen layout.
 ///
+/// Material Design 3 introduces 4 carousel layouts:
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:22]: Are all of these layouts supported by this widget? Maybe clarify in the docs what is supported right here.

File: packages/flutter/lib/src/material/carousel.dart [hunk 4]
```diff
@@ -15,31 +16,60 @@ import 'theme.dart';
...
+///  * Multi-browse: This layout shows at least one large, medium, and small
+/// carousel item at a time.
+///  * Uncontained (default): This layout show items that scroll to the edge of
+/// the container.
+///  * Hero: This layout shows at least one large and one small item at a time.
+///  * Full-screen: This layout shows one edge-to-edge large item at a time and
+/// scrolls vertically.
+///
 /// This widget supports uncontained carousel layout. It shows items that scroll
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:31]: Should this say: The default constructor implements the uncontained layout model.

File: packages/flutter/lib/src/material/carousel.dart [hunk 5]
```diff
@@ -15,31 +16,60 @@ import 'theme.dart';
...
+/// and [CarouselView.weighted] can build.
 ///
 /// ** See code in examples/api/lib/material/carousel/carousel.0.dart **
 /// {@end-tool}
 ///
 /// See also:
 ///
-///  * [CarouselController], which controls the first visible item in the carousel.
+///  * [CarouselController], which controls the first fully visible item in the
+/// view.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:72]:
```suggestion
///    view.
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 6]
```diff
@@ -58,11 +88,42 @@ class CarouselView extends StatefulWidget {
...
+    this.shape,
+    this.overlayColor,
+    this.itemSnapping = false,
+    this.shrinkExtent = 0.0,
+    this.controller,
+    this.scrollDirection = Axis.horizontal,
+    this.reverse = false,
+    this.allowFullyExpand = true,
+    this.onTap,
+    required this.layoutWeights,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:120]: Is it intentional that you could pass null in here or should null be disallowed with:

```suggestion
    required List<int> this.layoutWeights,
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 7]
```diff
@@ -140,16 +201,40 @@ class CarouselView extends StatefulWidget {
   /// Defaults to false.
   final bool reverse;
 
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:204]: nit: remove one of the blank lines

File: packages/flutter/lib/src/material/carousel.dart [hunk 8]
```diff
@@ -58,11 +88,42 @@ class CarouselView extends StatefulWidget {
     this.onTap,
     required this.itemExtent,
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:89]: Should developers be able to pass `null` in here, or should this be instead:

```suggestion
    required double this.itemExtent,
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 9]
```diff
@@ -159,13 +244,16 @@ class CarouselView extends StatefulWidget {
 }
 
 class _CarouselViewState extends State<CarouselView> {
-  late double _itemExtent;
+  double? _itemExtent;
+  List<int>? _weights;
   CarouselController? _internalController;
   CarouselController get _controller => widget.controller ?? _internalController!;
+  late bool allowFullyExpand;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:251]: Why do we need to store a copy of this in the state instead of just accessing it as `widget.allowFullyExpand` everywhere?

Comment (goderbauer) [reply packages/flutter/lib/src/material/carousel.dart:251]: If this property remains, it should likely be private:

```suggestion
  late bool _allowFullyExpand;
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 10]
```diff
@@ -159,13 +244,16 @@ class CarouselView extends StatefulWidget {
 }
 
 class _CarouselViewState extends State<CarouselView> {
-  late double _itemExtent;
+  double? _itemExtent;
+  List<int>? _weights;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:248]: Same question as below, can we just access `widget.layoutWeights` directly instead?

File: packages/flutter/lib/src/material/carousel.dart [hunk 11]
```diff
@@ -296,6 +388,51 @@ class _CarouselViewState extends State<CarouselView> {
...
+                        padding: effectivePadding,
+                        child: Material(
+                          clipBehavior: Clip.antiAlias,
+                          color: effectiveBackgroundColor,
+                          elevation: effectiveElevation,
+                          shape: effectiveShape,
+                          child: Stack(
+                            fit: StackFit.expand,
+                            children: <Widget>[
+                              widget.children.elementAt(index),
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:407]: Slightly simplified:

```suggestion
                              widget.children[index],
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 12]
```diff
@@ -140,16 +201,40 @@ class CarouselView extends StatefulWidget {
...
 
+
+  /// Whether the "squished" item is allowed to expand to the max size.
+  ///
+  /// If this is false, the layout of the carousel doesn't change. This is especially
+  /// useful when a weight list in [CarouselView.weighted] has a max item in the
+  /// middle and at least one small item on either side, such as `[1, 7, 1]`,
+  /// the first or the last item cannot expand to the max size. If this is true,
+  /// there will be some space before the first item or after the last item
+  /// coming so every items have a chance to be fully expanded.
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:212]:
```suggestion
  /// coming so every item has a chance to be fully expanded.
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 13]
```diff
@@ -140,16 +201,40 @@ class CarouselView extends StatefulWidget {
...
+  ///
+  /// Defaults to true.
+  final bool allowFullyExpand;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:215]: This name is not great :). Would something like `allowFirstAndLastToExpand` be better? Maybe other reviewers have better suggestions.

File: packages/flutter/lib/src/material/carousel.dart [hunk 14]
```diff
@@ -490,6 +627,506 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+      shrinkExtent: shrinkExtent,
+      weights: weights,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderSliverWeightedCarousel renderObject) {
+    renderObject.allowFullyExpand = allowFullyExpand;
+    renderObject.shrinkExtent = shrinkExtent;
+    renderObject.weights = weights;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:710]: Slightly more conventional:

```suggestion
    renderObject
      ..allowFullyExpand = allowFullyExpand
      ..shrinkExtent = shrinkExtent
      ..weights = weights;
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 15]
```diff
@@ -490,6 +627,506 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+    final int round = (constraints.scrollOffset / firstChildExtent).round();
+    if ((actual - round).abs() < precisionErrorTolerance) {
+      index = round;
+    } else {
+      index = actual.floor();
+    }
+    return allowFullyExpand ? index - smallerWeightCount : index;
+  }
+
+
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:843]: nit: remove one blank line

File: packages/flutter/lib/src/material/carousel.dart [hunk 16]
```diff
@@ -490,6 +627,506 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+      'This feature was deprecated after v3.20.0-7.0.pre.'
+    )
+    double itemExtent,
+  ) {
+    return childManager.childCount * maxChildExtent;
+  }
+
+  BoxConstraints _getChildConstraints(int index) {
+    double extent;
+    extent = itemExtentBuilder!(index, _currentLayoutDimensions)!;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:937]: Could be simplified:

```suggestion
    final double extent = itemExtentBuilder!(index, _currentLayoutDimensions)!;
```

File: packages/flutter/lib/src/material/carousel.dart (hunk 2 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:7]: I think it's the `max` getter from a List, like `weights.max` in this file

File: packages/flutter/lib/src/material/carousel.dart (hunk 3 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:22]: Updated! Thanks!

**Comment by bogdanenedoghitoiu** [comment]
Could you please add a `final ScrollBehavior? scrollBehavior`  property to the Carousel widget that will be passed to the Scrollable widget? 

It would be helpful in order to provide click and drag support for web or desktop apps, such as:
```dart
  scrollBehavior: ScrollConfiguration.of(context).copyWith(
    dragDevices: {
      PointerDeviceKind.touch,
      PointerDeviceKind.mouse,
    },
  ),
```

**Comment by Piinks** [comment]
> Could you please add a `final ScrollBehavior? 

ScrollBehavior is already inherited by the internal scrollable widget. You should be able to configure this by adding a ScrollConfiguration above the carousel. We usually only expose a scrollBehavior property if the widget is overriding it internally for a specific feature, making it inaccessible without exposing it as a property. I don't know that we need to do that in this case.

**Comment by bogdanenedoghitoiu** [comment]
> > Could you please add a `final ScrollBehavior?
> 
> ScrollBehavior is already inherited by the internal scrollable widget. You should be able to configure this by adding a ScrollConfiguration above the carousel. We usually only expose a scrollBehavior property if the widget is overriding it internally for a specific feature, making it inaccessible without exposing it as a property. I don't know that we need to do that in this case.

That makes sense, I will try it like this. Thank you for the reply!

File: packages/flutter/lib/src/material/carousel.dart [hunk 17]
```diff
@@ -56,13 +91,44 @@ class CarouselView extends StatefulWidget {
...
+       layoutWeights = null;
+
+/// Creates a scrollable list where the size of each child widget is dynamically
+/// determined by the provided [layoutWeights].
+///
+/// The `layoutWeights` parameter is required and defines the relative size
+/// proportions of each child widget.
+///
+/// When [allowFullyExpand] is set to `true`, each child can be expanded to its
+/// maximum size while scrolling. For example, with [layoutWeights] of `[1, 7, 1]`,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:106]: Is the maximum size dictated by the maximum weight?

File: packages/flutter/lib/src/material/carousel.dart [hunk 18]
```diff
@@ -56,13 +91,44 @@ class CarouselView extends StatefulWidget {
...
+  }) : allowFullyExpand = true,
+       layoutWeights = null;
+
+/// Creates a scrollable list where the size of each child widget is dynamically
+/// determined by the provided [layoutWeights].
+///
+/// The `layoutWeights` parameter is required and defines the relative size
+/// proportions of each child widget.
+///
+/// When [allowFullyExpand] is set to `true`, each child can be expanded to its
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:105]: If this max is the max weight (see other Q below), maybe calling this `consumeMaxWeight` is clearer?

File: packages/flutter/lib/src/material/carousel.dart [hunk 19]
```diff
@@ -56,13 +91,44 @@ class CarouselView extends StatefulWidget {
...
+    this.allowFullyExpand = true,
+    this.onTap,
+    required List<int> this.layoutWeights,
+    required this.children,
+  }) : itemExtent = null;
 
   /// The amount of space to surround each carousel item with.
   ///
-  /// Defaults to [EdgeInsets.all] of 4 pixels.
+  /// Defaults to EdgeInsets.all(4.0).
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:131]: Do we not want the breadcrumb?

File: packages/flutter/lib/src/material/carousel.dart [hunk 20]
```diff
@@ -105,7 +171,7 @@ class CarouselView extends StatefulWidget {
   /// adjusted to match this remaining space, ensuring a smooth size transition.
   ///
   /// Defaults to 0.0. Setting to 0.0 allows items to shrink/expand completely,
-  /// transitioning between 0.0 and the full [itemExtent]. In cases where the
+  /// transitioning between 0.0 and the full item size. In cases where the
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:174]: Is this no longer `itemExtent` because it could be based on a weighted size?

File: packages/flutter/lib/src/material/carousel.dart (hunk 13 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/carousel.dart:219]: Ah I wondered the same above. I thought maybe `consumeMaxWeight`?

File: packages/flutter/lib/src/material/carousel.dart [hunk 21]
```diff
@@ -140,16 +206,39 @@ class CarouselView extends StatefulWidget {
   /// Defaults to false.
   final bool reverse;
 
+  /// Whether the "squished" item is allowed to expand to the max size.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:209]:
```suggestion
  /// Whether the condensed item is allowed to expand to the max size.
```
WDYT?

Comment (Piinks) [reply packages/flutter/lib/src/material/carousel.dart:209]: Or collapsed? We use that term elsewhere in the framework.

File: packages/flutter/lib/src/material/carousel.dart [hunk 22]
```diff
@@ -140,16 +206,39 @@ class CarouselView extends StatefulWidget {
...
-  /// The item extent should not exceed the available space that the carousel
+  /// The item extent should not exceed the available space that the carousel view
   /// occupies to ensure at least one item is fully visible.
   ///
-  /// This must be non-null.
-  final double itemExtent;
+  /// This is required for [CarouselView]. In [CarouselView.weighted], this is null.
+  final double? itemExtent;
+
+  /// The weights that each visible child should occupy the viewport.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:232]:
```suggestion
  /// The weights that each visible child should occupy the in the viewport.
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 23]
```diff
@@ -140,16 +206,39 @@ class CarouselView extends StatefulWidget {
...
+  ///
+  /// The length of [layoutWeights] means how many items we want to lay out on
+  /// the viewport. For example, setting [layoutWeights] to `<int>[3,2,1]` means
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:235]:
```suggestion
  /// The length of [layoutWeights] represents how many items should be visible at a time
  /// in the viewport. For example, setting [layoutWeights] to `<int>[3,2,1]` means
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 24]
```diff
@@ -252,8 +351,8 @@ class _CarouselViewState extends State<CarouselView> {
               offset: position,
               clipBehavior: Clip.antiAlias,
               slivers: <Widget>[
-                _SliverFixedExtentCarousel(
-                  itemExtent: _itemExtent,
+                if (_itemExtent != null) _SliverFixedExtentCarousel(
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:354]: Rather than inlining the choice of sliver with if statements, can you create a private method that returns the right sliver here for better readability?

File: packages/flutter/lib/src/material/carousel.dart [hunk 25]
```diff
@@ -267,7 +366,52 @@ class _CarouselViewState extends State<CarouselView> {
...
                             children: <Widget>[
-                              widget.children.elementAt(index),
+                              widget.children[index],
+                              Material(
+                                color: Colors.transparent,
+                                child: InkWell(
+                                  onTap: () {
+                                    widget.onTap?.call(index);
+                                  },
+                                  overlayColor: widget.overlayColor ?? WidgetStateProperty.resolveWith((Set<WidgetState> states) {
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:376]: Can we pop this out to a method or getter as well?

File: packages/flutter/lib/src/material/carousel.dart [hunk 26]
```diff
@@ -490,6 +634,505 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+///
+/// _To learn more about slivers, see [CustomScrollView.slivers]._
+///
+/// This sliver arranges its children in a line along the main axis, starting
+/// at offset zero without gaps. Each child is constrained to its corresponding
+/// weight along the main axis and to the [SliverConstraints.crossAxisExtent]
+/// along the cross axis.
+///
+/// While scrolling, the extent (size) of each visible item changes dynamically
+/// based on the scrolling progress.  As the first visible item scrolls completely
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:648]:
```suggestion
/// based on the scrolling progress. As the first visible item scrolls completely
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 27]
```diff
@@ -490,6 +634,505 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+/// off-screen, the next item becomes the first visible item, and has the same
+/// size as the previously first item. The rest of the visible items maintain
+/// their relative layout.
+///
+/// For example, the layout weights is [1, 6, 1]. The length of [weights] array
+/// indicates three items should be visible at a time. The layout of these items
+/// would be:
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:655]:
```suggestion
/// For example, if the layout weights are [1, 6, 1], the length of [weights] array
/// indicates three items will be visible at a time. The layout of these items
/// would be:
```

File: packages/flutter/lib/src/material/carousel.dart [hunk 28]
```diff
@@ -490,6 +634,505 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+/// As item 0 scrolls off-screen, the extent of item 1 smoothly decreases from 600
+/// to 100. For instance, if item 0 is 30% off-screen, item 1 should have decreased
+/// its size to 30% of the difference from 600 to 100; its extent would be
+/// 600 - 0.3 * (600 - 100). Similarly, item 2's extent would increase from 100
+/// to 600, becoming 100 + 0.3 * (600 - 100).
+///
+/// As the initially visible items change size during scrolling, item 3 enters
+/// the view to fill the remaining space. Its extent starts at a minimum of
+/// [shrinkExtent] (or 0 if [shrinkExtent] is not provided) and gradually
+/// increases to match the extent of the last visible item (100 in this example).
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:671]: These docs are super valuable! But they will not be on https://api.flutter.dev/ since this is a private class. Should we move these docs up to the public class?

File: packages/flutter/lib/src/material/carousel.dart [hunk 29]
```diff
@@ -490,6 +634,505 @@ class _RenderSliverFixedExtentCarousel extends RenderSliverFixedExtentBoxAdaptor
...
+    final double extent = itemExtentBuilder!(index, _currentLayoutDimensions)!;
+    return constraints.asBoxConstraints(
+      minExtent: extent,
+      maxExtent: extent,
+    );
+  }
+
+  // This method is mostly the same as its parent class [RenderSliverFixedExtentList].
+  // The difference is when we allow some space before the leading items or after
+  // the trailing items with smaller weights, we leave extra scroll offset.
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:952]: Rather than copying most of this logic, can we add a spacing value/getter/method on the super class? If this is feasible, we could do it in a separate PR.

File: packages/flutter/test/material/carousel_test.dart [hunk 30]
```diff
@@ -26,12 +26,12 @@ void main() {
       ),
```
Comment (Piinks) [inline packages/flutter/test/material/carousel_test.dart:26]: Do we have tests for the new weighted constructor here that are horizontal, horizontal+reverse, vertical, and verical+reverse?

**Review by Piinks — COMMENTED** [review]
This is coming along so nicely! Mostly nits and a couple of questions. Thanks for your patience!

File: packages/flutter/lib/src/material/carousel.dart (hunk 13 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:219]: This one is shorter! I'll update it! But please let me know if there're any other preference😃

File: packages/flutter/lib/src/material/carousel.dart [hunk 31]
```diff
@@ -15,31 +16,65 @@ import 'theme.dart';
...
+/// [CarouselView.weighted] is used, then set the [layoutWeights] to only have
+/// one integer in the array.
+///
+/// In [CarouselView.weighted], weights are relative proportions. For example,
+/// if the layout weights is `[3, 2, 1]`, it means the first visible item occupies
+/// 3/6 of the viewport; the second visible item occupies 2/6 of the viewport;
+/// the last visible item occupies 1/6 of the viewport. As the carousel scrolls,
+/// the size of the latter one gradually changes to the size of the former one.
+/// As a result, when the first visible item is completely off-screen, the
+/// following items should stay the same layout as before. Using [CarouselView.weighted]
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:49]: Yes! Updated in the following commit:) Thanks!

File: packages/flutter/lib/src/material/carousel.dart [hunk 32]
```diff
@@ -15,31 +16,65 @@ import 'theme.dart';
...
+/// as indicated in [Carousel sepcs](https://m3.material.io/components/carousel/specs).
+///
+/// The [CarouselController] is used to control the
+/// [CarouselController.initialItem], which determines the first fully expanded
+/// item when the [CarouselView] or [CarouselView.weighted] is initially displayed.
+/// This is straightforward for [CarouselView] because each item in the view
+/// has fixed size. In [CarouselView.weighted], for instance, if the layout
+/// weights are `[1, 2, 3, 2, 1]` and the initial item is 4 (the fourth item), the
+/// view will display items 2, 3, 4, 5, and 6 with weights 1, 2, 3, 2 and 1
+/// respectively.
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:60]: Yes, it is right. When the initial item is 4, the 4th item should be in the middle initially in this case(1,2,3,2,1) and have the weight of 3. Does this make sense? Not very sure how to better explain this🤣

File: packages/flutter/lib/src/material/carousel.dart [hunk 33]
```diff
@@ -56,13 +91,44 @@ class CarouselView extends StatefulWidget {
...
+    this.shape,
+    this.overlayColor,
+    this.itemSnapping = false,
+    this.shrinkExtent = 0.0,
+    this.controller,
+    this.scrollDirection = Axis.horizontal,
+    this.reverse = false,
+    this.allowFullyExpand = true,
+    this.onTap,
+    required List<int> this.layoutWeights,
```
[the remark this answers is not in the document]
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:125]: Sounds good! This is not the language from M3 specs, just a name that I make up😄 Updated in following commits!

File: packages/flutter/lib/src/material/carousel.dart (hunk 17 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:106]: Yes. Updated.

**Comment by JinZr** [comment]
Dear all,

Thanks for your contribution! 

I'm really interested in this component and would like to ask whether this PR will be merged soon. 

Best regards
jin

File: packages/flutter/test/material/carousel_test.dart (hunk 30 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/carousel_test.dart:26]: Added some more tests for vertical, reverse, and itemSnapping combinations.

File: packages/flutter/lib/src/material/carousel.dart [hunk 34]
```diff
@@ -14,34 +14,75 @@ import 'ink_well.dart';
...
-/// viewport it occupies. This allows you to easily create layouts like
-/// multi-browse, hero, and full-screen.
+/// This code snippet shows how to get a vertical full-screen carousel by using
+/// [itemExtent] in [CarouselView].
+///
+/// ```dart
+/// Scaffold(
+///   body: CarouselView(
+///     scrollDirection: Axis.vertical,
+///     itemExtent: MediaQuery.sizeOf(context).height,
```
Comment (Piinks) [inline packages/flutter/lib/src/material/carousel.dart:55]: This would be the size of the device screen, not necessarily the size of the CarouselView. What if the carousel is smaller than the device screen?

File: packages/flutter/lib/src/material/carousel.dart (hunk 10 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/carousel.dart:291]: Oh! I wonder then if all we need are getters instead for the controller to be able to access these? Like:

```dart
List<int> get _weights => widget.flexWeights;
```

File: packages/flutter/lib/src/material/carousel.dart (hunk 28 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/carousel.dart:691]: Yeah, I think it is fine to include! :)

File: packages/flutter/lib/src/material/carousel.dart (hunk 10 above)
Comment (goderbauer) [reply packages/flutter/lib/src/material/carousel.dart:291]: > The whole purpose for this is when the initial item is given some value, we know how the initial view should look like.

Is this only used for initialization? Do we need to tell the controller (i.e. in didUpdateWidget) when these values have changed on the widget so the controller can redo its calculations with the new values?

(Instead copying I would also lean toward getters, the other question as to whether we need to tell the controller when these values change still stands, though)

File: packages/flutter/lib/src/material/carousel.dart [hunk 35]
```diff
@@ -175,6 +306,7 @@ class _CarouselViewState extends State<CarouselView> {
   @override
   void didChangeDependencies() {
     super.didChangeDependencies();
+    _consumeMaxWeight = widget.consumeMaxWeight;
     _itemExtent = widget.itemExtent;
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:310]: Why copy these in didChangeDependencies, but _weights is copied in initState?

File: packages/flutter/lib/src/material/carousel.dart [hunk 36]
```diff
@@ -730,13 +1426,27 @@ class CarouselController extends ScrollController {
   @override
   ScrollPosition createScrollPosition(ScrollPhysics physics, ScrollContext context, ScrollPosition? oldPosition) {
     assert(_carouselState != null);
-    final double itemExtent = _carouselState!._itemExtent;
+    final List<int>? weights = _carouselState!._weights;
+    final double? itemExtent = _carouselState!._itemExtent;
+    int expandedItem = initialItem;
+
+    if (weights != null && !_carouselState!._consumeMaxWeight) {
```
Comment (goderbauer) [inline packages/flutter/lib/src/material/carousel.dart:1433]: Related to the above question: What if the values read from _carouselState change after they have been read? Do we need to inform the controller and redo some of these calculations? Would also be good to add a test for that case.

File: packages/flutter/lib/src/material/carousel.dart (hunk 28 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart]: Updated!

File: packages/flutter/lib/src/material/carousel.dart (hunk 36 above)
Comment (Piinks) [reply packages/flutter/lib/src/material/carousel.dart:1436]: We've been chatting about this. I think what is needed here is to update the position, not the controller. The values provided to the controller are just passed to (and only used by) the position, which calculates what item is at what pixel, or what pixel is at which item based on those values.
So if they change here, we should update `_controller.position.VALUE` so it can make the right calculations based on the new info.

Converting the properties in the carousel scroll position to setters/getters will make this possible, with an update to the pixels when the value changes. (See _PagePosition.viewportFraction). And thanks Qun for mentioning PageView, looking at that really helped!

What do you all think?

Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:1436]: This makes sense to me:) Trying to add setters and getters for `itemExtent` and `flexWeights` in `_CarouselPosition` and setting they to new values when they are updated. Thanks for the suggestions!

Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart]: Just updated! `_flexWeights` and `_consumeMaxWeight` are changed to getters in `_CarouselViewState` class. `_itemExtent` should be constrained to be smaller than the viewport in the implementation, so I keep it as is. 

And when we update `flexWeights`, the layout will change to match the new weight list but the item that occupies the max weight stays the same. When we update `itemExtent`, the item main-axis extent will change but leading item also stays the same.

<details>
<summary>Demo 1 (Updating weights respects `initialItem`)</summary>

https://github.com/flutter/flutter/assets/36861262/f8b00fd0-da7b-47f0-b194-67bb45a13d49

</details>

<details>
<summary>Demo 2 (Updating weights respects the item that currently occupied the max weight)</summary>

https://github.com/flutter/flutter/assets/36861262/efdff60a-8a12-491b-a3e7-f69fd1e83177

</details>

File: packages/flutter/lib/src/material/carousel.dart (hunk 34 above)
Comment (QuncCccccc) [reply packages/flutter/lib/src/material/carousel.dart:55]: To make the `itemExtent` match the size of the carousel, we can use `double.infinity`. Just updated the sample code:)!

**Review by Piinks — APPROVED** [review]
Flutter_LGTM

**Comment by nguyenthanhphap1413** [comment]
Hi @QuncCccccc Hi, how can I get the index of the currently displayed items? Suppose I pass List<int>? _weights = [6,1], how can I get the index when I scroll to the positions?. Even the index of the position corresponding to 6 would be fine. Is there a way to get it or a formula to calculate it?

**Comment by bharathwajv** [comment]
can you please add options like this 

CarouselOptions(
              autoPlay: true,
              enlargeCenterPage: true,
              aspectRatio: 2.0,
            ),

**Comment by MarkOSullivan94** [comment]
What version of Flutter will have this?

**Comment by deminearchiver** [comment]
My wild guess is 3.26 stable, but it's already available on the main channel

**Comment by Piinks** [comment]
Hi everyone. The team (usually) does not follow up on comments in closed PRs. Please open a new issue for bugs, see [where's my commit](https://github.com/flutter/flutter/blob/master/docs/releases/Where's-my-commit.md) for what build tag is on a given PR, or see the help channel on [Discord](https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md) for questions.
